### PR TITLE
Linked headlines refactor

### PR DESCRIPTION
**Pull request by nileshgulia1**
- Use `<UniveralLink/>` in linked heading anchor since we are now using Links generated by `react-router-hashlink`.

### CODE REVIEW

**Comment by nileshgulia1** [comment]
This just needs a towncrier news item, can be added collectively in https://github.com/plone/volto/pull/3540.
@sneridagh @tiberiuichim  safe to merge?

**Comment by nileshgulia1** [comment]
@ksuess I remember you asked for moving the "slug icon" to the right just like mdn docs. Would you mind reviewing this PR then?

File: src/helpers/ScrollToTop/ScrollToTop.jsx
Comment (ksuess) [inline src/helpers/ScrollToTop/ScrollToTop.jsx:37]: I like isHash!

File: src/components/manage/Blocks/Title/View.jsx
Comment (ksuess) [inline src/components/manage/Blocks/Title/View.jsx:21]: Just for my understanding: What's the effect of useMemo here?

File: packages/volto-slate/src/editor/less/slate.less
Comment (ksuess) [inline packages/volto-slate/src/editor/less/slate.less:24]: I would prefer a link icon aligned and with same size of the headline.

Cosmetics: the link icon is rotated. Even if the shape of the icon is unusual, it's a global design decission.

File: packages/volto-slate/src/editor/render.jsx
Comment (ksuess) [inline packages/volto-slate/src/editor/render.jsx:178]: overriden by less CSS

**Review by ksuess — CHANGES_REQUESTED** [review]
This is great, I would take it as it is.
But an opt-out (no chain icon on hover) would be handsome.

File: src/components/manage/Blocks/Title/View.jsx
Comment (tiberiuichim) [reply src/components/manage/Blocks/Title/View.jsx:21]: @ksuess 

Let's say the code is like this:

```
const LinkedTitle = renderLinkElement('h1');
```

The renderLinkElements, whenever it is called, builds a new component. If for some reason the Title > View component is refreshed and runs again, the LinkedTitle would become a new component and react would unmount the old component `<LinkedTitle>` and create a new one instead. The `useMemo` is in fact a memoization, so that whenever the `View` component runs, it gets the same `LinkedTitle` component and so, when it's rendered, doesn't cause unnecessary unmount. In this case React would reuse the already mounted `<LinkedTitle>` and, if some prop changed (for example, the title was updated from the content), the component would receive the new prop (and, in class-based component terminology, would do a `componentDidUpdate`, while in function component, it would trigger any of its `React.useEffect()`.

File: packages/volto-slate/src/editor/less/slate.less
Comment (tiberiuichim) [reply packages/volto-slate/src/editor/less/slate.less:24]: @ksuess it's relatively difficult to make the link icon properly wrap with the text, so if it's multiple lines, it doesn't behave well. But my css knowledge may be limited, maybe there's a hack to do it

**Comment by nileshgulia1** [comment]
> But an opt-out (no chain icon on hover) would be handsome.

"no chain icon on hover" -  hm.. I don't see a reason why users will don't need it. @ksuess Could you give an example of sites where there is no link icon on hover but are anchors. Maybe I can just make that icon configurable? So a "#" in place of a chain icon.

File: src/components/manage/Blocks/Title/View.jsx
Comment (ksuess) [reply src/components/manage/Blocks/Title/View.jsx:21]: Got it. Thank you @tiberiuichim !

**Comment by sneridagh** [comment]
I also think it should be optional, not sure if "opt-in" or "opt-out"... I can certainly think it sites that do not need it (brochure, etc).

**Comment by nileshgulia1** [comment]
@sneridagh Sounds good to me.

**Comment by nileshgulia1** [comment]
@sneridagh Now that I started to implement this, I'm a bit puzzled having both "copy to clipboard" and "scroll to anchor" events triggering on clicking anchor icon. @albertcasado when an anchor is at the bottom of the page, clicking on it scrolls to that anchor. Could triggering a toast simultaneously annoy a user? 🤔

**Comment by nileshgulia1** [comment]
@sneridagh I have pushed the changelog. Would you mind reviewing it, please?

File: news/4287.feature
Comment (stevepiercy) [inline news/4287.feature:1]:
```suggestion
Added slug-based linked headlines in `volto-slate`. @tiberiuichim, @nileshgulia1
```

File: src/helpers/MessageLabels/MessageLabels.js
Comment (stevepiercy) [inline src/helpers/MessageLabels/MessageLabels.js:266]: If this message is targeted toward users instead of developers, I would suggest that the user-facing elements use the word "Link" instead of "URL". Thoughts?

Also is `UrlclipboardCopy` the correct casing? I don't know the convention to use when the leading characters are an acronym, and I would assume that the "c" in "clipboard" should be camelCase. Thus one of these may be more correct:

- `urlClipboardCopy`
- `URLClipboardCopy`

File: docs/source/configuration/volto-slate/configuration-settings.md
Comment (stevepiercy) [inline docs/source/configuration/volto-slate/configuration-settings.md:242]: We need to be consistent: heading != headline. A headline is usually a title, with some headings below. This might mean you would need to refactor the code as well.

```suggestion
(editor-configuration-slate-useLinkedHeading-label)=
```

File: docs/source/configuration/volto-slate/configuration-settings.md
Comment (stevepiercy) [inline docs/source/configuration/volto-slate/configuration-settings.md:246]:
```suggestion
The setting `slate.useLinkedHeadlines` controls whether `volto-slate` creates anchors for headings, such as `h1` and `h2`, in the editor.
The default setting is `true`.
You can opt out of this feature by setting its value to `false`.
```

Also possibly refactor the code and code sample, per previous comment.

**Review by stevepiercy — CHANGES_REQUESTED** [review]
I made a few suggestions and questions for your consideration.

File: docs/source/configuration/volto-slate/configuration-settings.md
Comment (nileshgulia1) [reply docs/source/configuration/volto-slate/configuration-settings.md:246]: Thanks @stevepiercy . I think "linked headings" would be much more precise than headlines.

Comment (nileshgulia1) [reply docs/source/configuration/volto-slate/configuration-settings.md]: Fixed in [de9e57f](https://github.com/plone/volto/pull/4287/commits/de9e57f3797e705a451351046a9688722a28500c)

File: docs/source/configuration/volto-slate/configuration-settings.md
Comment (nileshgulia1) [reply docs/source/configuration/volto-slate/configuration-settings.md]: [de9e57f](https://github.com/plone/volto/pull/4287/commits/de9e57f3797e705a451351046a9688722a28500c)

File: src/helpers/MessageLabels/MessageLabels.js
Comment (nileshgulia1) [reply src/helpers/MessageLabels/MessageLabels.js:266]: [de9e57f](https://github.com/plone/volto/pull/4287/commits/de9e57f3797e705a451351046a9688722a28500c)

File: news/4287.feature
Comment (nileshgulia1) [reply news/4287.feature]: Fixed in [de9e57f](https://github.com/plone/volto/pull/4287/commits/de9e57f3797e705a451351046a9688722a28500c)

**Comment by nileshgulia1** [comment]
@stevepiercy I resolved your comments in [de9e57f](https://github.com/plone/volto/pull/4287/commits/de9e57f3797e705a451351046a9688722a28500c). Could you please have a look? Thanks!

**Comment by nileshgulia1** [comment]
> @sneridagh Now that I started to implement this, I'm a bit puzzled having both "copy to clipboard" and "scroll to anchor" events triggering on clicking anchor icon. @albertcasado when an anchor is at the bottom of the page, clicking on it scrolls to that anchor. Could triggering a toast simultaneously annoy a user? 🤔

@stevepiercy @sneridagh Anyways, what do you think of this opinion, As a user? Can we disable toast and instead show a copied "SVG" near the anchor.

**Comment by stevepiercy** [comment]
> > @sneridagh Now that I started to implement this, I'm a bit puzzled having both "copy to clipboard" and "scroll to anchor" events triggering on clicking anchor icon. @albertcasado when an anchor is at the bottom of the page, clicking on it scrolls to that anchor. Could triggering a toast simultaneously annoy a user? 🤔
> 
> @stevepiercy @sneridagh Anyways, what do you think of this opinion, As a user? Can we disable toast and instead show a copied "SVG" near the anchor.
> 
> **Current scenario**
> 
>  Screen.Recording.2023-06-09.at.8.57.08.PM.mov

For code blocks in docs, we have this interface, which is nice. It has both the universal icon for copy, plus translatable text. You can play with it at the permalink:

https://6.docs.plone.org/volto/configuration/how-to.html#extending-configuration-in-a-project

https://github.com/plone/volto/assets/102112/5fff227c-7ce9-47d8-bfae-e4c465abda5d

I think the toast at the bottom is too much and disconnected from the click on large screens.

**Review by stevepiercy — APPROVED** [review]
I approve docs and naming conventions. A maintainer should review source code, too. Thank you!

**Comment by nileshgulia1** [comment]
Thank you @stevepiercy .

File: packages/volto-slate/src/editor/render.jsx
Comment (davisagli) [inline packages/volto-slate/src/editor/render.jsx:186]: Is the wrapper really needed? It looks to me like it would be possible to put the margin and vertical-align on the link itself if it uses `display: inline-block` instead of `display: contents`. It would be nice to keep the markup simple and semantic, especially since this is part of the public UI.

File: src/components/manage/Blocks/ToC/variations/DefaultTocRenderer.jsx
Comment (davisagli) [inline src/components/manage/Blocks/ToC/variations/DefaultTocRenderer.jsx:23]: I expected that clicking the link in the ToC would update the location to include the hash (so that it's easy to copy from the address bar), but it did not.

File: packages/volto-slate/src/editor/less/slate.less
Comment (davisagli) [inline packages/volto-slate/src/editor/less/slate.less:19]: It's possible to construct a heading with just the right (wrong?) length so that the anchor svg flows to the next line and takes up additional vertical space. To avoid that, I'd suggest adding `position: absolute;` so that the anchor never affects positioning! It means it might flow into the margin in this edge case, but that's probably not a big problem.

File: src/components/manage/Blocks/Title/View.jsx
Comment (davisagli) [inline src/components/manage/Blocks/Title/View.jsx:30]: There might be some other blocks that need a similar adjustment, like the listing block which has a headline field.

**Review by davisagli — CHANGES_REQUESTED** [review]
I took a look and I like it! I do have a few suggestions...

Should we close https://github.com/plone/volto/pull/3540 and edit this PR to set the target branch to master? I don't think merging it into https://github.com/plone/volto/pull/3540 first has any particular benefit.

**Comment by davisagli** [comment]
> Anyways, what do you think of this opinion, As a user? Can we disable toast and instead show a copied "SVG" near the anchor.

@nileshgulia1 I think it would be nice to have it closer to where the user clicks, but I am not sure how to guarantee that there is space to show it. The toast is okay and has the benefit of (hopefully) following existing practices for a11y rather than needing to make sure something new is accessible...I would keep the toast for now and come back to it as a future improvement.

File: packages/volto-slate/src/editor/less/slate.less
Comment (nileshgulia1) [reply packages/volto-slate/src/editor/less/slate.less]: I was having problems with the alignment of anchor with title. For example if text is "aligned"(center,left) or if font is too big...etc. Anyway, I pushed your suggestions in [c33f634](https://github.com/plone/volto/pull/4287/commits/c33f634e32bcd7eb56ae80d36dd5258f2aeb101e)

File: src/components/manage/Blocks/Title/View.jsx
Comment (nileshgulia1) [reply src/components/manage/Blocks/Title/View.jsx:30]: adapted listing block headline : [c33f634](https://github.com/plone/volto/pull/4287/commits/c33f634e32bcd7eb56ae80d36dd5258f2aeb101e)

File: src/components/manage/Blocks/ToC/variations/DefaultTocRenderer.jsx
Comment (nileshgulia1) [reply src/components/manage/Blocks/ToC/variations/DefaultTocRenderer.jsx]: [c33f634](https://github.com/plone/volto/pull/4287/commits/c33f634e32bcd7eb56ae80d36dd5258f2aeb101e)

File: packages/volto-slate/src/editor/render.jsx
Comment (nileshgulia1) [reply packages/volto-slate/src/editor/render.jsx]: Removed in [c33f634](https://github.com/plone/volto/pull/4287/commits/c33f634e32bcd7eb56ae80d36dd5258f2aeb101e)

File: packages/volto-slate/src/hooks/useClipboard.jsx
Comment (tiberiuichim) [inline packages/volto-slate/src/hooks/useClipboard.jsx:1]: Please move this to volto's main codebase.

File: packages/volto-slate/src/editor/less/slate.less
Comment (davisagli) [reply packages/volto-slate/src/editor/less/slate.less]: It's certainly possible there are still edge cases, but it looks okay to me now.

**Review by davisagli — APPROVED** [review]
LGTM now

**Review by sneridagh — APPROVED** [review]

**Pull request merged**
